source/extensions/tracers/opentelemetry/samplers/trace_id_ratio_based/trace_id_ratio_based_sampler.cc

```diff
@@ -0,0 +1,102 @@
+#include "source/extensions/tracers/opentelemetry/samplers/trace_id_ratio_based/trace_id_ratio_based_sampler.h"
+
+#include <cstdint>
+#include <cstring>
+#include <iostream>
+#include <memory>
+#include <sstream>
+#include <string>
+
+#include "envoy/type/v3/percent.pb.h"
+
+#include "source/common/common/safe_memcpy.h"
+#include "source/extensions/tracers/opentelemetry/span_context.h"
+
+namespace {
+/**
+ * Converts a ratio in [0, 1] to a threshold in [0, MILLION].
+ */
+uint64_t ratioToThreshold(double ratio) noexcept {
+  const uint64_t MAX_VALUE = Envoy::ProtobufPercentHelper::fractionalPercentDenominatorToInt(
+      envoy::type::v3::FractionalPercent::MILLION);
+
+  if (ratio <= 0.0) {
+    return 0;
+  }
+  if (ratio >= 1.0) {
+    return MAX_VALUE;
+  }
+
+  return static_cast<uint64_t>(ratio * static_cast<double>(MAX_VALUE));
+}
```

Comment:
I actually didn't get this. Although `double` is find as ratio for me. But if we use fractionl percent helper here, why not use the `envoy.type.v3.FractionalPercent` directly and use `evaluateFractionalPercent` to get the decsion?

Reply:
Hi @wbpcode,
would you mind taking another look at the code?